### ISSUE TITLE
FactAttribute with CancellationToken parameter

### ISSUE THREAD

**Issue by sharwell** [xunit/xunit#345]
Currently a test with the following form can be created.

```csharp
[Fact]
public async Task TestSomething()
{
}
```

It would be extremely helpful if the following signature was supported as well.

```csharp
[Fact]
public async Task TestSomething(CancellationToken cancellationToken)
{
}
```

In the future, the provided `CancellationToken` might support being cancelled if the user manually cancels a test run before it is complete, or perhaps based on other criteria like a limit on the total number of test failures that can occur before cancelling a parallel test run.

For now, it would be perfectly acceptable to simply pass `CancellationToken.None` for the parameter if/when it is encountered.

**Comment by AArnott** [comment]
I would add as an additional use case: cancellation when the test times out. I know xunit2 doesn't support the TimeoutAttribute or FactAttribute(Timeout) parameter, but it would be nice to support if nothing else by way of a passed in CancellationToken to the test method.

**Comment by sharwell** [comment]
@AArnott I meant to make that a separate issue so we can expedite support for just these signatures. Would that work for you?

**Comment by AArnott** [comment]
Yes, that's fine. I guess the new issue would depend on this one.

**Comment by bradwilson** [comment]
Out of curiosity, from where would this cancellation token be manufactured?

**Comment by bradwilson** [comment]
We could _sort of_ implement this today, but I'd wager you'd sometimes be unhappy with the result.

A little background: on most platforms, the unit tests are loaded and executed in a separate app domain from the one that launched the runner (the console EXE, MSBUILD, etc.). When a user wants to cancel tests, this information is sitting in the runner. Today, the only mechanism for communicating that intent is once the execution engine on the other side has sent a message back to the runner (to indicate the status of the tests, etc.). There is no mechanism today for having the runner call into the execution engine to say "stop". Cancellation tokens are not supported across app domains, so that isn't a candidate.

What this means is, effectively, cancelling a test means waiting for _something_ to finish so that we can tell the system to stop running more things. If you were blocked on a single thing that seemed to have run away, your intent to cancel would be sitting waiting for a message to come from the other side, so as to communicate that intention.

There are... some hacks that I could imagine putting in here. The simplest one would be to add some kind of heartbeat message from the execution engine simply for the purpose of letting the runner call back and say "Yes, please cancel". I'm going to have to think about ways that this could be accomplished without breaking forward compatibility from 2.0 runners to 2.1 tests.

**Comment by AArnott** [comment]
Yes, what you propose was about what I was thinking. I look forward to hearing what you come up with.

Sent from my Windows Phone

-----Original Message-----

**Comment by bradwilson** [comment]
After discussion, we've decided not to do this feature.

**Comment by AArnott** [comment]
Where was the discussion? I'd like to read a transcript if its available (on slack perhaps?)
I guess I'll have to write another extension then, because I have a code base of tests that depend on timeouts in mstest since product failures so often result in test hangs in this particular code base.

**Comment by bradwilson** [comment]
There is no transcript. Jim and I discuss the issues with each other when we get together.

**Comment by AArnott** [comment]
OK. Can you tell me, in the event that I implement it, whether the CancellationTokenSource given to a test runner gets canceled if the user tries to cancel the test run from within Test Explorer in VS?

**Comment by bradwilson** [comment]
Yes. Example: https://github.com/xunit/xunit/blob/3732f359e529859a9a5049fba6d17ca603116c11/src/xunit.execution/Sdk/Frameworks/Runners/TestCollectionRunner.cs#L94-L95

Unless we've missed some places, the test infrastructure should end up cancelling the CTS when they've received a "stop running" directive from the user (as a result of reporting some status message back to the runner).

You still have the caveat that those directives only come across the runner/execution boundary when something has happened that results in the reporting of a message from execution => runner. There's really no way to overcome that without an architectural change in the core system itself. There is, therefore, no way to predict when the cancellation will take place, and further, if you're in a deadlocked situation with nothing else running, no actual way to effect the cancellation at all.

**Comment by AArnott** [comment]
OK, thanks.

**Comment by niemyjski** [comment]
+1

**Comment by AArnott** [comment]
I'm still interested in the timeout-based cancellationtoken being passed to the test method. When it was decided against doing this, is there a design that contradicts this idea, or would you accept a pull request?

**Comment by Piedone** [comment]
Did you do anything on this front @AArnott @sharwell ? Both straight away cancellation token handling (to cancel long-running UI tests) and cancellation based on the total number of test failures would be quite useful for us.

**Comment by AArnott** [comment]
@Piedone This is how I solve the problem in all of my test projects:

https://github.com/AArnott/Nerdbank.Streams/blob/42d6ce345050b32c1a7854ae861a7eee7f90b737/src/Nerdbank.Streams.Tests/TestBase.cs#L51

This is in a base class from which I derive all my test classes.

**Comment by AArnott** [comment]
@piedone When the test times out, it throws OperationCanceledException from the line that got "stuck". The test fails because it threw. I'm not sure what you're looking for in terms of "graceful", but for my purposes, a timeout is a test failure, and that's how xunit interprets it.

**Comment by Piedone** [comment]
I mean, canceling a given test execution from the runner. But I see, thanks.

**Comment by AArnott** [comment]
The only clean way a runner can terminate a test execution is to kill the test process. xunit can't do that, AFAIK. So coding each test carefully to cancel when called upon is the next best thing.

**Comment by Piedone** [comment]
There is no way to make this cancellation on-demand (i.e. not based on timeout) though, right? Like my test (i.e. a single method) is running for three minutes and I want to cancel it mid-flight but let it handle its cancellation. So, it's not possible right now what this issue asked for and one has to do workarounds like [this one](https://github.com/xunit/xunit/issues/2008#issuecomment-588321441).

**Comment by AArnott** [comment]
If you want to cancel it with something other than a simple timer, why not just set up a CancellationTokenSource within your test method and call Cancel on it when you want to? That's certainly possible today. What are you hoping xunit would add to improve this?

**Comment by AArnott** [comment]
Oh! I see. You want the runner's shutdown/abort sequence to signal a CancellationToken that the test can consume. I misunderstood.

**Comment by alx9r** [comment]
@Piedone I'm also interested in graceful handling of user-invoked cancellation during test execution.  (I _think_ that's what you were driving at earlier.)  Were you able to confirm whether the following is the case?

>...there's nothing to tie the cancellation of the token source to, since there's no way to get notified about test execution being canceled from the runner...

Did you settle on some workaround or approximation of graceful user-invoked cancellation?

**Comment by Piedone** [comment]
Yes, I still think that's the case and no, I didn't find a solution at all. I actually tried the cancel keypress workaround but that event wasn't raised for me at all (might be specific to `dotnet test`, what I'm only interested in). So no good news, unfortunately.

**Comment by snerte** [comment]
Even though there currently are technical limitations and this might therefore not be feasible, it would be a great if we could add a cancellationtoken to tests that is triggered when -for example- a user cancels a test.

**Comment by jzabroski** [comment]
@bradwilson Given Microsoft has adopted Xunit so liberally, and it is by far the most used testing framework today, is there any chance of making this possible?

**Comment by JobaDiniz** [comment]
I have a test that will run 100 tasks in the background, but if the test is stopped (I click stop button on VS), then I want all of the running task to cancel.
If I receive the `CancellationToken` from the underlying test system (`xUnit`) and I can propagate it to my tasks.

**Comment by paulegradie** [comment]
Interesting discussion - Just coming across this now while writing a simple xUnit test that 'needs' a cancellation token as well. I'm not sure why this isn't able to imlemented in xUnit, but I can testify that its very useful behavior to have in a test runner depending on the runner. 

[JobaDiniz](https://github.com/JobaDiniz) - I actually chose to implement CancellationToken injection in the [Sailfish Test Adapter](https://github.com/paulegradie/Sailfish).

You'll notice in the getting started example you can request a cancellation token and link that to your invocations. When you cancel the test run (when executed in the IDE) - the internal cancellation token updates the cancellation requested property along with setting the test runner internal state to 'cancelled'. 

(Of course - when using Sailfish in a console app, you can provide your own cancellation token to the main Run method, or let Sailfish create one for you.)

**Comment by bradwilson** [comment]
`CancellationToken` is part of `TestContext` in v3: https://github.com/xunit/xunit/blob/4d1f2e5d4ac9260487d0a8f35a2d045388021b33/src/xunit.v3.core/TestContext.cs#L32-L37

**Comment by dmytro-gokun** [comment]
@bradwilson Thanks for the hint! Would not it be nice if this was automatically passed to test methods which have CancellationToken as their last parameter? :)

Until this is implemented, we can declare something like this in our (base) test class:

```
protected CancellationToken TestCancellationToken => TextContext.Current.CancellationToken;
```

**Comment by bradwilson** [comment]
We have never supported "injection" into method values, and are not likely to add this. The only injection we have historically done is constructor injection, which you can do in v3 for `ITestContextAccessor` (if you'd prefer to avoid statics): https://github.com/xunit/xunit/blob/4d1f2e5d4ac9260487d0a8f35a2d045388021b33/src/xunit.v3.core.tests/Acceptance/Xunit3AcceptanceTests.cs#L722-L738

The text context has a lot of potential values in it. We will not be supporting injection of these values; you must attempt to retrieve them via the context.
